packages/editor/src/lib/editor/Editor.ts

```diff
@@ -4573,31 +4573,38 @@ export class Editor extends EventEmitter<TLEventMap> {
 	 * @public
 	 */
 	@computed getCurrentPageShapesSorted(): TLShape[] {
-		// todo: consider making into a function call that includes options for selected-only, rendering, etc.
-		// todo: consider making a derivation or something, or merging with rendering shapes
 		const shapes = new Set(this.getCurrentPageShapes().sort(sortByIndex))
-
-		const results: TLShape[] = []
-
-		function pushShapeWithDescendants(shape: TLShape): void {
-			results.push(shape)
-			shapes.delete(shape)
-
-			shapes.forEach((otherShape) => {
```

Comment:
This was the problematic part. For each shape we added we went through all the other shapes to find it's children. Now we have a map of parent child relationships which gets generated on the first pass through the shapes.

Reply:
damn, good one!